src/main/java/net/neoforged/neoforge/transfer/resources/ResourceStack.java

```diff
@@ -0,0 +1,211 @@
+        if (amount == this.amount) return this;
+        if (amount == 0 || resource.isEmpty()) return resource().getEmptyResourceStackInstance();
+        return ResourceStack.of(resource, amount);
+    }
+
+    /**
+     * @param amount Amount to shrink by. Must be non-negative.
+     * @return A new immutable instance of this resource stack with an updated amount decreased by the specified {@code amount}.
+     *         If the newAmount is 0 or the resource is empty, then the EMPTY instance for the resource will be returned.
+     *
+     * @throws IllegalArgumentException when {@code amount} is negative.
+     *
+     * @see #withAmount(int)
+     */
+    public ResourceStack<T> shrink(int amount) {
+        TransferPreconditions.checkNonNegative(amount);
+        return withAmount(Math.max(this.amount - amount, 0));
+    }
+
+    /**
+     * @param amount Amount to grow by. Must be non-negative.
+     *
+     * @return A new immutable instance of this resource stack with an amount increased by the specified {@code amount}.
+     *         If the resource was already empty, then the EMPTY instance will be returned instead
+     *
+     * @throws IllegalArgumentException when {@code amount} is negative.
+     */
+    public ResourceStack<T> grow(int amount) {
+        TransferPreconditions.checkNonNegative(amount);
+        return withAmount(this.amount + amount);
+    }
+
+    /**
+     * @return A new this instance with an updated resource should it have changed, otherwise it returns itself.
+     */
+    public ResourceStack<T> with(UnaryOperator<T> operator) {
+        return with(operator, amount);
+    }
+
+    /**
+     * @param amount the amount the new stack should be.
+     * @return A new this instance with an updated resource should it have changed, otherwise it returns itself.
+     */
+    public ResourceStack<T> with(UnaryOperator<T> operator, int amount) {
+        T result = operator.apply(resource);
+        if (amount == 0) return result.getEmptyResourceStackInstance();
+
+        if (result.equals(resource) && amount == amount()) return this;
+        return ResourceStack.of(result, amount);
+    }
+
+    /**
+     * Checks if this is empty, meaning that the amount is zero
+     * or that the resource is {@link IResource#isEmpty() empty}.
+     *
+     * @return {@code true} if empty
+     */
+    public boolean isEmpty() {
+        return ResourceHandlerUtil.isEmpty(resource(), amount());
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        ResourceStack<?> that = (ResourceStack<?>) obj;
+        return amount == that.amount && resource.equals(that.resource);
+    }
+
+    @Override
+    public int hashCode() {
+        return 31 * resource.hashCode() + Integer.hashCode(amount);
+    }
+
+    @Override
+    public String toString() {
+        return "%s(%d)".formatted(resource, amount);
```

Comment:
You either need to pass a locale, or just do concatenation. Personally I think using `formatted` is overkill for this